chalk-solve/src/clauses.rs

```diff
@@ -18,6 +18,45 @@ mod env_elaborator;
 mod generalize;
 pub mod program_clauses;
 
+// yields the types "contained" in `app_ty`
+fn consistuent_types<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    app_ty: &ApplicationTy<I>,
+) -> Vec<Ty<I>> {
+    let interner = db.interner();
+    match app_ty.name {
+        TypeName::Adt(adt_id) => {
+            let adt_datum = &db.adt_datum(adt_id);
+            let adt_datum_bound = adt_datum.binders.substitute(interner, &app_ty.substitution);
+            adt_datum_bound
+                .variants
+                .into_iter()
+                .map(|variant| variant.fields.into_iter())
+                .flatten()
```

Comment:
`.flat_map(|variant| variant.fields.into_iter())`

Reply:
Oh, didn't even know about that one. Thanks!